### ISSUE TITLE
[ttnn/eltwise]: bfloat16 add and subtract return 0.0 for Inf - Inf where float32 returns NaN, the operand dtype selects the compute engine

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#52141]
**Component / Area:** ttnn/eltwise/binary_ng

**Issue Type:** Bad Outputs

## Observed

`ttnn.subtract` and `ttnn.add` return `0.0` for `Inf - Inf` when the operands are `bfloat16`, and `nan` when they are `float32`. Same expression, same shapes, only the dtype differs.

Measured on Blackhole P300 at `b2e77906b9`, a clean build with nothing applied. Every tile is filled with the value, and every result is uniform across all 1024 elements:

| expression | IEEE / torch | `ttnn` bfloat16 | `ttnn` float32 |
|---|---|---|---|
| `add(+Inf, -Inf)` | NaN | **`0.0`** | `nan` |
| `subtract(+Inf, +Inf)` | NaN | **`0.0`** | `nan` |
| `multiply(+Inf, 0)` | NaN | **`0.0`** | `nan` |
| `divide(+Inf, +Inf)` | NaN | **`+Inf`** | `nan` |
| `add(+Inf, 1)` | +Inf | `+Inf` | `+Inf` |
| `subtract(+Inf, 1)` | +Inf | `+Inf` | `+Inf` |
| `multiply(+Inf, 2)` | +Inf | `+Inf` | `+Inf` |
| `add(-Inf, -Inf)` | -Inf | `-Inf` | `-Inf` |
| `reciprocal(+Inf)` | +0 | `0.0` | `0.0` |
| `neg(+Inf)` | -Inf | `-Inf` | `-Inf` |
| `clone(+Inf)` | +Inf | `+Inf` | `+Inf` |

The last four rows are the control: everything IEEE defines as finite-or-infinite comes out right in both dtypes. Only the invalid operations differ, and only in bfloat16. `to_layout` on a tile carrying `0x7F80` and `0xFF80` returns both unchanged, so the packer is not involved and this reproduces on plain `main`.

The `multiply` and `divide` rows each have a different identified cause and are not what this issue is about; see Scope.

## Expected

`nan`. IEEE 754 calls `Inf - Inf` an invalid operation whose result is NaN, and `bfloat16` represents NaN perfectly well, so nothing about the type forces this.

**The dtype selects the engine.** `ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp:19-31`:

```cpp
bool is_binary_sfpu_op(BinaryOpType val, DataType a, DataType b, bool fast_and_approximate_mode = false) {
    switch (val) {
        case ADD:
        case SUB:
        ...
        case RSUB: return a == b && (a == FLOAT32 || a == INT32 || a == UINT32 || a == UINT16);
```

`BFLOAT16` is absent from that list, so bfloat16 `add`, `subtract` and `rsub` go to the FPU. The value becomes `operation_attributes.is_sfpu` and selects the compute kernel at `binary_ng_program_factory.cpp:890-891`. It is computed from the input dtypes only.

**The FPU has no exponent-all-ones.** `tt-isa-documentation`, `WormholeB0/TensixTile/TensixCoprocessor/FloatBitPatterns.md`, gives three interpretations of each bit pattern: IEEE 754, Vector Unit (SFPU), and Matrix Unit (FPU). For exponent all ones the FPU column reads `(1 + Mant/2^7) * 2^(Exp-127)`, an ordinary very large finite number. So on the FPU `Inf - Inf` is (huge finite) minus (the same huge finite), and the answer is zero.

**It is the engine, not the accumulator width.** `fp32_dest_acc_en` is enabled whenever an operand or the output is `float32` (`binary_ng_program_factory.cpp:1178-1183`), so the float32 column above also differs in Dest width. Asking for a `float32` output from `bfloat16` inputs widens Dest and leaves the dispatch alone, which separates the two: `add(+Inf, -Inf)` and `subtract(+Inf, +Inf)` are still `0.0`.

**Confirmed from the other side.** Adding `BFLOAT16` to that `return` and rebuilding changes both of them to a NaN, and nothing else in the table moves.

Note that the `return` is shared. `ADD`, `SUB` and `RSUB` fall through to it together with `LOGICAL_AND`, `LOGICAL_OR`, `LOGICAL_XOR` and `SQUARED_DIFFERENCE`, so editing it in place moves seven operations between engines rather than three. That is fine as a diagnostic, but a change intended to fix only `add` and `subtract` has to split the labels first.

The same question has already been answered once, for a different op, in the opposite direction. #35077 moved `bfloat16` multiply from the FPU to the SFPU in 2026-01 to fix its rounding, and added an explicit clamp so that `0 * Inf` would keep returning `0` as it had on the FPU. So the healing is deliberate for multiply and incidental for add, and neither is written down.

## Steps to reproduce

```python
import torch, ttnn

d = ttnn.open_device(device_id=0)

def bf16(bits):                      # a 32x32 bfloat16 tile filled with `bits`
    r = torch.full((1, 1, 32, 32), bits - (1 << 16) if bits >= 0x8000 else bits, dtype=torch.int16)
    return ttnn.from_torch(r.view(torch.bfloat16), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=d)

def f32(bits):
    r = torch.full((1, 1, 32, 32), bits - (1 << 32) if bits >= 0x80000000 else bits, dtype=torch.int32)
    return ttnn.from_torch(r.view(torch.float32), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=d)

print(ttnn.to_torch(ttnn.subtract(bf16(0x7F80), bf16(0x7F80))).flatten()[0])          # bfloat16 Inf - Inf
print(ttnn.to_torch(ttnn.subtract(f32(0x7F800000), f32(0x7F800000))).flatten()[0])    # float32  Inf - Inf
print(ttnn.to_torch(ttnn.add(bf16(0x7F80), bf16(0xFF80))).flatten()[0])               # bfloat16 Inf + (-Inf)
print(ttnn.to_torch(ttnn.add(f32(0x7F800000), f32(0xFF800000))).flatten()[0])         # float32  Inf + (-Inf)
```

```
tensor(0.)
tensor(nan)
tensor(0.)
tensor(nan)
```

Build a fresh input for each measurement and read each result once: several ttnn ops deallocate their input, so reusing one device tensor across ops reads a freed buffer.

## Scope

`add`, `subtract` and `rsub` with both operands `bfloat16`, on any shape and either memory config. The dispatch line is checked once per op from the input dtypes, so nothing about the tensor changes the answer.

Not this defect, though they appear in the table above:

- **`multiply(+Inf, 0)`** is an explicit clamp in `calculate_sfpu_binary_mul`, `v_if(in0 == 0 || in1 == 0) { result = 0.0f; }`, added by #35077 and commented "To match FPU behaviour for bfloat16 multiplication". Deliberate, and in the opposite direction to this issue.
- **`divide(+Inf, +Inf)`** is #51557, the packer. The SFPU computes the NaN correctly. `sfpu_reciprocal_iter` opens with "sfpi::approx_recip(x) will return ±0 for x = ±inf", so `r` is `+0`; the Newton-Raphson step then computes `t = x*y - 2.0f` which is `Inf*0 - 2.0f`, and the source notes "when x=0 and y=infinity (and vice versa), t=+NaN regardless of the operand signs". The `v_if(t < 0)` guard exists precisely so a NaN `t` skips the refinement and the `±0` survives. `calculate_sfpu_binary_div` then computes `in0 * r` = `Inf * 0` = NaN, and the Markstein refinement is skipped for bfloat16 anyway. So a NaN reaches the store and `+Inf` comes out of the bfloat16 tensor. The rounding in that path is not responsible: the same NaN-to-`+Inf` substitution appears on the SFPU add path identically with and without `float32_to_bf16_rne` applied.

## Why no test catches it

`tech_reports/Handling_Special_Value/special_values.md` gives one table per engine, and they differ in exactly the two rows this issue is about:

| Operation | FPU table | SFPU table |
|---|---|---|
| `±Inf × ±Inf`, `±finite × ±Inf`, `Inf + Inf`, `-Inf – Inf` | listed | listed |
| `0 × Inf` | **absent** | `NaN` |
| `+Inf - Inf` | **absent** | `NaN` |

with the text "the operations not listed in the tables above will treat `NaN`/`Inf` numbers just as any other ordinary number". Both statements are accurate. What is missing is which engine a given ttnn op with a given dtype runs on, which is the only thing a caller of `ttnn.subtract` can act on. No test asserts that routing rule, because it is not written anywhere to assert against.

There is an established in-tree narrative for bfloat16 special values, and it is about a different mechanism: `pytest.xfail("NaN is packed as inf for ttnn.bfloat16")` appears in `test_div_ops.py`, `test_where.py`, `test_unary.py` and others, and `test_pow.py:167` has "Due to HW limitations for bfloat16 dtype, NaN value gets packed as inf." Those are correctly attributed to the packer (#51557). The routing behaviour here is a second, independent mechanism that produces a different wrong answer (`0.0`, so no NaN is ever created for the packer to destroy), and no workaround in the tree mentions it.

## What fixing it costs

I built the routing change so the decision does not have to be made blind. It is small but not a one-liner: the `case` labels have to be split so `LOGICAL_AND`, `LOGICAL_OR`, `LOGICAL_XOR` and `SQUARED_DIFFERENCE` keep their current engine, and `get_sfpu_init_fn` in `binary_ng_utils.cpp` has to opt the bfloat16 path into rounding, which depends on #51060.

**Accuracy: no cost, once the SFPU add path rounds its output.** As it stands the SFPU add and subtract truncate their bfloat16 result, so the routing change on its own makes ordinary values worse (mean absolute error 1.716e-3 to 2.326e-3). That is a separate defect, already tracked as #49944 with a PR in #51060. With both, 65536 random bfloat16 pairs:

| | mean abs error | max abs error |
|---|---|---|
| torch bfloat16 add, i.e. correctly rounded | 1.720e-3 | 7.812e-3 |
| bfloat16 add on the FPU, today | 1.722e-3 | 8.301e-3 |
| bfloat16 add on the SFPU, with the rounding fix | **1.720e-3** | **7.812e-3** |

So the routing change ends up marginally more accurate than the path it replaces, not less. It just cannot be made alone. #51060's per-call-site `ROUND_NEAREST` flag is the shape that would let `binary_ng` opt in without affecting anything else.

**Throughput, from the device profiler**, median cycles per core per invocation at 1350 MHz:

| shape, memory | | unpack `TRISC_0` | math `TRISC_1` | pack `TRISC_2` | whole op |
|---|---|---|---|---|---|
| 512x512, L1 | `add` on the FPU, today | 4882 | **178** | 2960 | 5612 |
| 512x512, L1 | on the SFPU, with the rounding fix | 4941 | **5054** | 3337 | 5860 (**+4.4%**) |
| 2048x2048, DRAM | `add` on the FPU, today | 83526 | 78633 | 82064 | 84170 |
| 2048x2048, DRAM | on the SFPU, with the rounding fix | 83436 | 83750 | 82362 | 84464 (+0.3%) |

The math thread costs 28x more and the op costs 4.4%, because the math thread was not the critical path: the unpacker was. In DRAM the difference disappears into the memory traffic. `ttnn.multiply`'s `fast_and_approximate_mode`, which switches the same two engines on an unmodified build, gives the same picture from the other direction: math 227 against 5182 cycles, whole op 5646 against 5936, +5.1%.

The LLK microbenchmark that compares `Elwadd` against `SfpuElwadd` reports 3.4x for the isolated kernel. That number does not transfer to the op, because it makes the math thread the critical path and `binary_ng` does not.

## Impact

`Inf - Inf` is the shape of the standard softmax stabilisation, `x - max(x)`. If a logit has already overflowed to `+Inf`, IEEE says that entry becomes NaN and the divergence is visible. In bfloat16 it becomes `0.0`, `exp` of which is `1.0`, so that entry silently receives the weight of the maximum and the run continues with a plausible-looking result. A NaN check after the op never fires, because no NaN is ever produced.

This is a reachability argument, not a measured one: no model was traced, and reaching it requires an overflow to have already happened.

What I am asking for, in order of usefulness:

1. **A decision** on whether bfloat16 `add` and `subtract` should keep FPU special-value semantics. #35077 answered the same question for multiply by keeping them deliberately, so "yes, for the same reason" is a perfectly good answer. What is missing is that the answer exists anywhere a caller can find it.
2. **The routing rule written down**: which engine runs which op for which dtype. The hardware half is already in `special_values.md`; the dispatch half is not, and neither is the fact that bfloat16 multiply deliberately does not implement the `0 × Inf` row.
3. If the answer is "keep the FPU", **a note in `special_values.md` that bfloat16 add and subtract heal invalid operations to `0.0`**, so the behaviour is discoverable without reading the dispatch.

## Environment

- Blackhole P300, device 0
- tt-metal `b2e77906b9`, clean build, nothing applied
- sfpi 7.67.0

## Not established

- **Whether the FPU routing for bfloat16 add is a deliberate throughput choice or an accident.** The dispatch line was grown by three integer-support PRs (#21116, #27290, #33768), none of which mentions bfloat16 or special values, which suggests the latter, but no one has said so.
- **Wormhole.** Everything here is Blackhole P300. The dispatch code is architecture independent and the ISA document quoted above is the Wormhole one, so the same behaviour is expected, but it is not measured.

## Related

- #51557 — the packer destroying a bfloat16 NaN. Different mechanism, and not a prerequisite: infinities survive the pack, which is why everything above reproduces on plain `main`. It is why the NaN version of the same comparison is invisible today, and it is what turns the `divide(+Inf, +Inf)` row into `+Inf`.
- #49944 / #51060 — the SFPU add path's bfloat16 rounding. A prerequisite for the routing change discussed above, not part of this issue.
- #51470 — `maximum`/`minimum`/`clamp`/`hardtanh` dropping a NaN through `SFPSWAP`, an SFPU permutation rather than FPU routing.
- #51498 (`clamp` substituting a finite sentinel) and #33885 (`inf == inf` false in sfpi branches) are adjacent but separate.
- #27218, closed in 2025-11, asked why comparison operators do not recognise inf/nan and was answered with "the hardware does not return it automatically, software has to interpret it". The same theme at the dispatch layer rather than the kernel layer.

**Comment by kinginu** [comment]
One more data point for "Why no test catches it".

#11243, merged 2024-09, resolved a set of nan/inf mismatches by moving the goldens rather than the device: *"Based on our discussion we will be replacing the nan / inf output from the torch with the numbers to match the TT output. For this, I have used a `torch.nan_to_num` function."* It covers unary and composite ops, not the binary `add`/`subtract` this issue is about, so it is not the same defect — but it is the same response.

Together with the `pytest.xfail("NaN is packed as inf for ttnn.bfloat16")` family, which attributes correctly to the packer (#51557), that makes two occasions on which bfloat16 special-value divergence has been absorbed into the test suite without the underlying mechanism being recorded. Engine routing is a third mechanism, and it is the one that yields `0.0` rather than an `inf` or a substituted golden — so it leaves no trace in either workaround.
